packages/nx/src/project-graph/project-graph.ts

```diff
@@ -50,7 +50,72 @@ import { DelayedSpinner } from '../utils/delayed-spinner';
 export function readCachedProjectGraph(
   minimumComputedAt?: number
 ): ProjectGraph {
-  const projectGraphCache = readProjectGraphCache(minimumComputedAt);
+  let projectGraphCache = readProjectGraphCache(minimumComputedAt);
+
+  if (!projectGraphCache && !IS_WASM) {
+    // Check if another process is currently building the graph
+    const lockPath = join(workspaceDataDirectory, 'project-graph.lock');
+
+    // If a lock exists, wait synchronously for the other process to finish building
+    if (fileExists(lockPath)) {
+      const maxWaitTime = 120000; // 120 seconds max wait
+      const pollInterval = 200; // 200ms between checks
+      const startTime = Date.now();
+
+      logger.verbose(
+        `[readCachedProjectGraph] Waiting for another process to build the project graph...`
+      );
+
+      // Create a shared buffer for Atomics.wait()
+      const sharedBuffer = new SharedArrayBuffer(4);
+      const sharedArray = new Int32Array(sharedBuffer);
+
+      while (!projectGraphCache && Date.now() - startTime < maxWaitTime) {
+        // Use Atomics.wait for efficient synchronous sleep
+        // This blocks the thread without consuming CPU
+        Atomics.wait(sharedArray, 0, 0, pollInterval);
+
+        // Log progress occasionally
+        const elapsedMs = Date.now() - startTime;
+        if (elapsedMs > 0 && elapsedMs % 5000 < pollInterval) {
+          // Every 5 seconds
+          const elapsedSeconds = Math.round(elapsedMs / 1000);
+          logger.verbose(
+            `[readCachedProjectGraph] Still waiting... (${elapsedSeconds}s elapsed)`
+          );
+        }
+
+        // Check if graph is now available
+        projectGraphCache = readProjectGraphCache(minimumComputedAt);
+
+        // Check if lock still exists
+        if (!fileExists(lockPath)) {
```

Comment:
The presence of the lock file doesn't actually indicate the graph building. After the first graph is created, the file will always remain on disk. We place an OS read write  lock on the file which is the actual indicator